### PR TITLE
#163783849 Add cascading effect on soft-delete feature

### PR DESCRIPTION
**Pull request by RonKbS**
#### What does this PR do?
The PR enables the deletion of related data if a delete mutation is used

#### Description of Task to be completed?
When deletions are done, related data is not included in the deletion/archiving. As an example, deleting a location will leave the rooms for that particular location still available. This PR enables this to be carried out automatically.

#### How should this be manually tested?
Pull the branch and run a deletion query such as the one of locations below.

``` 
mutation {
  deleteLocation(locationId: 1) {
    location {
      name
    }
  }
}
```

Any offices and rooms related to the location will also be deleted/archived.

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#163977386](https://www.pivotaltracker.com/story/show/163977386)

#### Screenshots (if appropriate)
N/A

### CODE REVIEW

File: utilities/utility.py [hunk 1]
```diff
@@ -14,6 +16,30 @@ def update_entity_fields(entity, **kwargs):
...
+    def receive_after_update(mapper, connection, target):
+        @event.listens_for(db_session, "after_flush", once=True)
+        def receive_after_flush(session, context):
+            if target.state == 'archived':
+                child_model = getattr(api, child)
+                # Check for resource name is due to the improper
+                # naming of folder room_resource and class Resource
+                child_object = getattr(
+                    child_model.models,
+                    'Resource' if child.capitalize() == 'Room_resource' else child.capitalize()  # noqa: E501
```
Comment (caveinn) [inline utilities/utility.py]: I think you  can remove then noqa by having the if on statement on different lines or having a child.capitilize() assigned to a variable with a shorter name.

File: utilities/utility.py [hunk 2]
```diff
@@ -14,6 +16,30 @@ def update_entity_fields(entity, **kwargs):
...
+                )
+                parent_s_id = getattr(child_object, parent_id)
```
Comment (caveinn) [inline utilities/utility.py]: I think parents_id would be more intuitive the s seems misplaced in the above name.

File: api/room/models.py [hunk 3]
```diff
@@ -59,3 +59,11 @@ def receive_after_flush(session, context):
...
             raise GraphQLError("Room calendar Id is invalid")
         pass
+
+
+cascade_soft_delete(
+    Room, 'room_resource', 'room_id'
+)
+cascade_soft_delete(
+    Room, 'events', 'room_id'
+)
```
Comment (bevkololi) [inline api/room/models.py]: Kindly include devices and responses in soft-deletion of room responses.

File: utilities/utility.py (hunk 2 above)
Comment (RonKbS) [reply utilities/utility.py]: parents_id would refer to more than one parent. 
parent_s_id replaces ' in parent's hence its use here.

File: api/room/models.py (hunk 3 above)
Comment (RonKbS) [reply api/room/models.py]: Responses and devices have no state column in their models so calling this function on them would not work.

As discussed in the backend-sync, stories are to be raised to fix this.

File: utilities/utility.py [hunk 4]
```diff
@@ -14,6 +16,30 @@ def update_entity_fields(entity, **kwargs):
     return entity
 
 
+def cascade_soft_delete(parent_model, child, parent_id):
+    @event.listens_for(parent_model, 'after_update')
```
Comment (michael-basweti) [inline utilities/utility.py]: @RonKbS is there a way we can have the two wrappers in one overall function, this way we can reduce complexity by reducing the number of functions.

Comment (RonKbS) [reply utilities/utility.py]: I have just pushed changes with one each decorator in its own function and this should solve the complexity issue @michael-basweti.

Thanks for the feedback.

File: utilities/utility.py [hunk 5]
```diff
@@ -14,6 +16,35 @@ def update_entity_fields(entity, **kwargs):
...
+    def receive_after_update(mapper, connection, target):
+        listen_for_after_flush_event(target, child, parent_id)
+
+
+def listen_for_after_flush_event(target, child, parent_id):
+    @event.listens_for(db_session, "after_flush", once=True)
+    def receive_after_flush(session, context):
+        if target.state == 'archived':
+            child_model = getattr(api, child)
+            # Check for resource name is due to the improper
```
Comment (Bonifase) [inline utilities/utility.py]: KIndly include your comments immediately after the function definition. Use docstrings https://www.python.org/dev/peps/pep-0008/#block-comments

File: utilities/utility.py [hunk 6]
```diff
@@ -14,6 +16,35 @@ def update_entity_fields(entity, **kwargs):
...
+            # naming of folder room_resource and class Resource
+            child_object = getattr(
+                child_model.models,
+                'Resource' if child.capitalize() == 'Room_resource'
+                else child.capitalize()
+            )
+            parent_s_id = getattr(child_object, parent_id)
```
Comment (Bonifase) [inline utilities/utility.py]: Did you mean `parents_id`? If not so then use a more descriptive variable naming

**Comment by RonKbS** [comment]
The recommended changes have been implemented @Bonifase.

File: utilities/utility.py [hunk 7]
```diff
@@ -14,6 +16,41 @@ def update_entity_fields(entity, **kwargs):
...
+        The check statement on line 42 caters for the difference
+        in naming of room_resource folder and Resource class
+            :param session:
+            :param context:
+        """
+        if target.state == 'archived':
+            child_model = getattr(api, child)
+            child_object = getattr(
+                child_model.models,
+                'Resource' if child.capitalize() == 'Room_resource'
```
Comment (nakatuddesuzan) [inline utilities/utility.py]: Hello @RonKbS , good work. help me understand more about this code by explaining why there is a need to use the capitalize method here. Thanks

Comment (RonKbS) [reply utilities/utility.py]: @nakatuddesuzan, the method needs to run queries on an object like `api.room.models.Room` and to make it generic one string is passed in (`room`) that is then capitalized to `Room`

**Comment by RonKbS** [comment]
> Great work on the cascade I've tested this and it works @RonKbS, Kindly fix the merge conflicts.

Conflicts have been resolved @RachaelNantale.

**Comment by dnuwa** [comment]
tested and works fine. good work @RonKbS

**Pull request merged**